registry/remote/repository.go

```diff
@@ -126,6 +128,33 @@ func NewRepository(reference string) (*Repository, error) {
 	}, nil
 }
 
+// newRepositoryWithOptions returns a Repository with the given reference and options.
+// RepositoryOptions are part of the Registry struct and may set defaults for the Registry.
+// RepositoryOptions share the same struct definition as Repository and therfore contain
+// unexported state that we must avoid copying around to different Repositories.
+// To handle this we explicitly copy only the fields that we want to reproduce.
+func newRepositoryWithOptions(ref registry.Reference, opts *RepositoryOptions) (*Repository, error) {
+	if err := ref.ValidateRepository(); err != nil {
+		return nil, err
+	}
+	return &Repository{
+		Client:               opts.Client,
+		Reference:            ref,
+		PlainHTTP:            opts.PlainHTTP,
+		ManifestMediaTypes:   cloneSlice(opts.ManifestMediaTypes),
+		TagListPageSize:      opts.TagListPageSize,
+		ReferrerListPageSize: opts.ReferrerListPageSize,
+		MaxMetadataBytes:     opts.MaxMetadataBytes,
+	}, nil
+}
+
+func cloneSlice[S ~[]E, E any](s S) S {
```

Comment:
nit: Can we have some documentation and unit tests for this new method?

Reply:
Done